src/providers/vmware/amd64.rs

```diff
@@ -39,8 +63,155 @@ impl VmwareProvider {
     fn fetch_guestinfo(erpc: &mut vmw_backdoor::EnhancedChan, key: &str) -> Result<Option<String>> {
         let guestinfo = erpc
             .get_guestinfo(key.as_bytes())
-            .context("failed to retrieve network kargs")?
+            .with_context(|| format!("failed to retrieve guestinfo for {}", key))?
             .map(|bytes| String::from_utf8_lossy(&bytes).into_owned());
         Ok(guestinfo)
     }
+
+    pub fn parse_netplan_config(&self) -> Result<Option<String>> {
+        if let Some(metadata) = &self.guestinfo_metadata {
+            // We need to parse the netplan config to remove the non-Netplan keys.
+            // The data can either be JSON or YAML, but since JSON is a subset of
+            // YAML we don't need to try serde_json::from_str here.
+            let netplan_config_unfiltered: serde_json::Value = serde_yaml::from_str(metadata)?;
+            // Only the "network" key is allowed to be present.
+            // We use the json! macro but this is only about creating a serde value::Value,
+            // even though its name sounds like it would create JSON.
+            let netplan_config_filtered = json!({ "network": netplan_config_unfiltered.get("network").ok_or(anyhow!("no 'network' key found"))?});
+            let netplan_config = serde_yaml::to_string(&netplan_config_filtered)?;
+            Ok(Some(netplan_config))
+        } else {
+            Ok(None)
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_from_metadata(metadata: String) -> Result<Self> {
+        Ok(Self {
+            guestinfo_net_kargs: None,
+            guestinfo_metadata: Some(metadata),
+        })
+    }
+}
+
+fn parse_metadata(
+    guestinfo_metadata_encoding: Option<String>,
+    guestinfo_metadata_raw: Option<String>,
+) -> Result<Option<String>> {
```

Comment:
I think we could avoid some of the indentation here by leveraging pattern matching more. E.g. we could reflow this function as:

```rust
match (
	guestinfo_metadata_encoding_val.as_deref(),
	guestinfo_metadata_raw
) {
	(Some("base64" | "b64")) => {
		...
		Ok(Some(...))
	...
	(Some("") | None, guestinfo_metadata_raw) => Ok(...),
	(Some(enc), _) => Err(anyhow!("unknown guestinfo.metadata.encoding {enc}")),
}
```

Reply:
Thanks